tensorflow/core/platform/ctstring_internal.h

```diff
@@ -162,6 +162,14 @@ static inline size_t TF_TString_ToInternalSizeT(size_t size,
 #else   // TF_TSTRING_LITTLE_ENDIAN
   // 0xFF000000 or 0xFF00000000000000 depending on platform
   static const size_t mask = ~((~(size_t)0) >> 8);
+#if defined(__s390x__)
+  if (type == TF_TSTR_OFFSET) {
```

Comment:
Thank you for finding and reporting this bug!
My current understanding is that `TF_TString_ToInternalSizeT` should only be used with "large" and "view" representations that use `size_t size`. The "offset" representation uses "uint32_t size". My thought is that the issue should be solved by creating a `TF_TString_ToInternalUint32T` function (which uses `uint32_t` and not `size_t`) and have the test use this new function.

Reply:
To clarify, the bug here is in the test; ctstring_test.cc is the only user of `TF_TString_ToInternalSizeT(..., TF_TSTR_OFFSET);` which as @SeeForTwo  mentioned is an invalid use.  This test was added after the original implementation.

The test should not be using internal functions for generating offset types.  Offset types were written as a placeholder for a future potential need for memory map-able tstrings, and as such, generation of offset types was not implemented, as they would need to be done with a series of tstrings in mind in order to generate an offset table and its associated packed contiguous data segement.  Generating a single OFFSET type doesn't have much use.

To resolve this either: 1) remove the test, or since OFFSET types are expected to be little-endian u32, 2) construct a byte sequence by hand for size and offset.

Reply:
Good summary.

> 2) construct a byte sequence by hand for size and offset.

This seems reasonable. While Offset types are currently a placeholder, some code for them exists and at can be exercised, so I see advantage in keeping and fixing the test.

Any other thoughts?

Reply:
I have no preference.

Reply:
Looking closer at the existing test, there are several issues with it. @namrata-ibm please replace the OffsetType test with the following and try it out on your big-endian arch.   If there are no issues, please update your pull request to include:

```
 TEST(TF_CTStringTest, OffsetType) {
    {
      uint8_t str[] = "test";
      constexpr size_t str_size = sizeof(str) / sizeof(str[0]);
   
      uint8_t buf[sizeof(TF_TString) + str_size];
   
      memcpy(buf + sizeof(TF_TString), str, str_size);
   
      TF_TString *offsets = (TF_TString *)buf;
      TF_TString_Init(offsets);
   
      offsets[0].u.offset.size = TF_le32toh(str_size << 2 | TF_TSTR_OFFSET);
      offsets[0].u.offset.offset = TF_le32toh(sizeof(TF_TString));
      offsets[0].u.offset.count = TF_le32toh(1);
                                                                                                                                                                                                                   
      EXPECT_EQ(str_size, TF_TString_GetSize(offsets));
      EXPECT_EQ(TF_TSTR_OFFSET, TF_TString_GetType(offsets));
      EXPECT_EQ(0, ::memcmp(str, TF_TString_GetDataPointer(offsets), str_size));
    }
  }
```

Reply:
Note that we are 'technically' doing 'TF_htole32', but a macro with that name doesn't currently exist.  It would be the same exact function though. Feel free to add one in ctstring_internal.h or keep as is (and maybe make a comment)

Reply:
Thank you @cantonios @SeeForTwo @gharibian for your valuable comments. 
@gharibian I tried above test on BE, however `EXPECT_EQ(0, ::memcmp(str, TF_TString_GetDataPointer(offsets), str_size));` this check produces a crash on BE.

Reply:
@namrata-ibm, thank you for your patience! I would defer to @gharibian, but I think that in TF_TString_GetDataPointer the current:
```
    case TF_TSTR_OFFSET:
      return (const char *)str + str->u.offset.offset;  // NOLINT
```

should be changed to use TF_le32toh:

```
    case TF_TSTR_OFFSET:
      return (const char *)str + TF_le32toh(str->u.offset.offset);  // NOLINT
```

Reply:
@SeeForTwo Yes, that works on LE and BE.

Reply:
That looks good, thanks for catching that @SeeForTwo

Reply:
> Yes, that works on LE and BE.

@namrata-ibm, wonderful! If you are happy with the working software that you have, could you please update the PR when you have time?  (My apologies if you've already updated and I'm just confused by how to see it in github.)

Reply:
@SeeForTwo Updated the PR, Thank you!